File: src/main/java/com/example/blogging_platform/writer/WriterService.java
package com.example.blogging_platform.writer;

import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/** Service bean for writer */
@Service
public class WriterService {

    /** Repository bean for writer */
    private final WriterRepository writerRepository;

    /** Constructor */
    public WriterService(WriterRepository writerRepository) {
        this.writerRepository = writerRepository;
    }

    /**
     * Fetch all writers
     * 
     * @return All writers
     */
    public List<Writer> getAllWriters() {
        return writerRepository.findAll();
    }

    /**
     * Fetch a writer with single ID
     * 
     * @param id Writer Id to fetch
     * @return Writer with requested ID. If no such writer, the result is empty.
     */
    public Optional<Writer> getWriterById(Long id) {
        return writerRepository.findById(id);
    }

    /**
     * Create a new writer
     * 
     * @param writer New writer
     * @return Newly created writer
     */
    public Writer createWriter(Writer writer) {
        return writerRepository.save(writer);
    }

    /**
     * Update a writer with new information
     * 
     * @param id Writer Id to update
     * @param updatedWriter New information to be updated
     * @return Updated writer
     * @throws java.lang.RuntimeException No writer is found with given ID
     */
    public Writer updateWriter(Long id, Writer updatedWriter) {
        return writerRepository.findById(id)
                .map(existingWriter -> {
                    existingWriter.editWriterInfo(updatedWriter);
                    return writerRepository.save(existingWriter);
                })
                .orElseThrow(() -> new RuntimeException("Writer not found with id " + id));
    }

    /**
     * Delete a writer with single ID
     * 
     * @param id Writer ID to delete
     * @throws java.lang.RuntimeException No writer is found with given ID
     */
    public void deleteWriter(Long id) {
        if (!writerRepository.existsById(id)) {
            throw new RuntimeException("Writer not found with id " + id);
        }
        writerRepository.deleteById(id);
    }
}

File: src/main/java/com/example/blogging_platform/writer/WriterService.kt
package com.example.blogging_platform.writer

import org.springframework.stereotype.Service
import java.util.*
import java.util.function.Function

@Service
class WriterService(private val writerRepository: WriterRepository) {
    val allWriters: List<Writer?>
        get() = writerRepository.findAll()

    fun getWriterById(id: Long): Optional<Writer?> {
        return writerRepository.findById(id)
    }

    fun createWriter(writer: Writer): Writer {
        return writerRepository.save(writer)
    }

    fun updateWriter(id: Long, updatedWriter: Writer): Writer {
        return writerRepository.findById(id)
            .map<Writer>(Function { existingWriter: Writer? ->
                existingWriter?.editWriterInfo(updatedWriter)
                existingWriter?.let { writerRepository.save(it) }
            })
            .orElseThrow {
                RuntimeException(
                    "Writer not found with id $id"
                )
            }
    }

    fun deleteWriter(id: Long) {
        if (!writerRepository.existsById(id)) {
            throw RuntimeException("Writer not found with id $id")
        }
        writerRepository.deleteById(id)
    }
}
